v2/api/network/customizations/load_balancer_extension.go

```diff
@@ -0,0 +1,171 @@
+func fuzzySetInboundNatRules(lb genruntime.ARMResourceSpec, inboundNatRules []genruntime.ARMResourceSpec) (err error) {
+	if len(inboundNatRules) == 0 {
+		// Nothing to do
+		return nil
+	}
+
+	defer func() {
+		if x := recover(); x != nil {
+			err = errors.Errorf("caught panic: %s", x)
+		}
+	}()
+
+	// Here be dragons
+	lbValue := reflect.ValueOf(lb)
+	lbValue = reflect.Indirect(lbValue)
+	if (lbValue == reflect.Value{}) {
+		return errors.Errorf("cannot assign to nil loadbalancer")
+	}
+
+	propertiesField := lbValue.FieldByName("Properties")
+	if (propertiesField == reflect.Value{}) {
+		return errors.Errorf("couldn't find properties field on loadbalancer")
+	}
+
+	propertiesValue := reflect.Indirect(propertiesField)
+	if (propertiesValue == reflect.Value{}) {
+		// If the properties field is nil, we must construct an entirely new properties and assign it here
+		temp := reflect.New(propertiesField.Type().Elem())
+		propertiesField.Set(temp)
+		propertiesValue = reflect.Indirect(temp)
+	}
+
+	inboundNatRulesField := propertiesValue.FieldByName("InboundNatRules")
+	if (inboundNatRulesField == reflect.Value{}) {
+		return errors.Errorf("couldn't find InboundNatRules field on LoadBalancer")
+	}
+
+	if inboundNatRulesField.Type().Kind() != reflect.Slice {
+		return errors.Errorf("inboundNatRules field on LoadBalancer was not of kind Slice")
+	}
+
+	elemType := inboundNatRulesField.Type().Elem()
+	inboundNatRuleSlice := reflect.MakeSlice(inboundNatRulesField.Type(), 0, 0)
+
+	for _, inboundNatRule := range inboundNatRules {
+		embeddedInboundNatRule := reflect.New(elemType)
+		err := fuzzySetInboundNatRule(inboundNatRule, embeddedInboundNatRule)
+		if err != nil {
+			return err
+		}
+
+		inboundNatRuleSlice = reflect.Append(inboundNatRuleSlice, reflect.Indirect(embeddedInboundNatRule))
+	}
+
+	// Now do the assignment. We do it differently here, as we need to make sure to retain current/updated/deleted InboundNatRules present on the LoadBalancer.
+	inboundNatRulesField.Set(reflect.AppendSlice(inboundNatRulesField, inboundNatRuleSlice))
+
+	return nil
+}
+
+func fuzzySetInboundNatRule(InboundNatRule genruntime.ARMResourceSpec, embeddedInboundNatRule reflect.Value) error {
+	inboundNatRuleJSON, err := json.Marshal(InboundNatRule)
+	if err != nil {
+		return errors.Wrapf(err, "failed to marshal inboundNatRule json")
+	}
+
+	err = json.Unmarshal(inboundNatRuleJSON, embeddedInboundNatRule.Interface())
+	if err != nil {
+		return errors.Wrapf(err, "failed to unmarshal inboundNatRule JSON")
+	}
+
+	// Safety check that these are actually the same:
+	// We can't use reflect.DeepEqual because the types are not the same.
+	embeddedInboundNatRuleJSON, err := json.Marshal(embeddedInboundNatRule.Interface())
+	if err != nil {
+		return errors.Wrap(err, "unable to check that embedded inboundNatRule is the same as inboundNatRule")
+	}
+	if string(embeddedInboundNatRuleJSON) != string(inboundNatRuleJSON) {
```

Comment:
Does JSON serialization have a guarantee of ordering? I suspect this may be fragile.

Reply:
This is the same logic we've been using for other resource extensions(vnet, routetable) as well and noticed no serialization issues. Can you think of a better way to do this?

Reply:
According to this [stackoverflow post](https://stackoverflow.com/questions/44755089/does-serialized-content-strictly-follow-the-order-in-definition-use-encoding-jso), which references the Go JSON serialization code, struct order will be consistent.

That's an implementation detail though and is technically subject to change.

Map ordering is not guaranteed, so if the structs had maps we could be at risk here. Since we can't know for sure that the structs don't have maps, I think in general I agree with Bevan that this is not ideal.

_Possibly_ what we could do here is:
1. Serialize these types to JSON
2. Deserialize them into a generic `map[string]interface{}` (map of maps)
3. use `reflect.DeepEqual` on those.

While the actual JSON representations of the types might differ because of ordering, I _believe_ that `reflect.DeepEqual` is not ordering sensitive for maps and so using that would be safe.

As above it might make sense to merge this as-is (given it's a bug we have across all the impls right now) and fix it across all of them together for consistency.

Reply:
if we do decide to fix this but in a different PR, let's file a bug for it so we don't forget. Make sure to mention RouteTable and VNETs there too.